Implement escaping of webhook params

When serving webhooks with user provided data, such as the deploy webhook, we want the received webhook parameters to be properly escaped.

This can be done by providing each argument to the `printf` utility with a '%q' specifier. 
```sh
printf %q "$ARG1"
```
This will print a properly escaped value of `$ARG1` by escaping all special characters such as quotes, slashes, etc in the `$ARG1` variable.

The webhook image we are running is a based on Alpine and does not contain most GNU tools, therefore the `printf` utility in the image does not support `%q` specifier.

https://www.shellcheck.net/wiki/SC3050

The `webhook/hooks` directory contains scripts for all the webhooks, it also houses utility scripts, prefixed with a `.` such as `.authenticate.sh`, consider creating a `.escape.sh` script to achieve the functionality of `printf %q` to be used in the webhook scripts.

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -314,12 +314,14 @@ jobs:
             # cache node modules using the same key as restore.
           key: ${{ steps.node_modules-cache-restore.outputs.cache-primary-key }}
 
-  Starchart-Docker:
+  Docker-Build-Deploy:
     # We'll only build and push when landing on main, not for PRs
     if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
     runs-on: ubuntu-latest
     # Don't bother unless everything else passes
     needs: [Build, Prettier, Unit-Tests, E2E-Tests, Dockerfile-Lint, ESLint, Type-Check]
+    env:
+      IMAGE: ghcr.io/DevelopingSpace/starchart
     permissions:
       contents: read
       packages: write
@@ -327,18 +329,11 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v3
 
+      - name: Define GITHUB_SHORT_SHA
+        run: echo "GITHUB_SHA_SHORT=$(echo $GITHUB_SHA | cut -c 1-6)" >> $GITHUB_ENV
+
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v2
-
-      - name: Extract Docker metadata from git info
-        id: meta
-        uses: docker/metadata-action@v4
-        with:
-          images: ghcr.io/DevelopingSpace/starchart
-          flavor: latest=true
-          tags: |
-            type=ref,event=branch
-            type=sha
 
       - name: Log in to the GitHub Container Registry
         uses: docker/login-action@v2
@@ -354,5 +349,16 @@ jobs:
           build-args: |
             SAML_IDP_METADATA_PATH=./config/idp-metadata-staging.xml
           push: true
-          tags: ${{ steps.meta.outputs.tags }}
-          labels: ${{ steps.meta.outputs.labels }}
+          tags: |
+            ${{ env.IMAGE }}:sha-${{ env.GITHUB_SHA_SHORT }}, \
+            ${{ env.IMAGE }}:main, \
+            ${{ env.IMAGE }}:latest
+
+      - name: Deploy to Staging via Webhook
+        uses: navied/secure-actions-webhook@0.2.0
+        env:
+          HMAC_SECRET: ${{ secrets.WEBHOOK_STAGING_SECRET }}
+        with:
+          url: https://mycustomdomain-dev.senecacollege.ca/hooks/deploy
+          hmacSecret: ${{ env.HMAC_SECRET }}
+          data: '{"tag": "sha-${{ env.GITHUB_SHA_SHORT }}"}'

--- a/DEPLOY.md
+++ b/DEPLOY.md
@@ -5,7 +5,7 @@ Startchart depends on a number of external services, including:
 1. MySQL
 2. Amazon Route53
 3. An SMTP Server
-4. A Redis Server
+4. A Redis Server (run via Docker)
 5. Let's Encrypt
 6. SAML2 IdP (e.g., Azure Active Directory)
 
@@ -17,20 +17,20 @@ A number of environment variables and Docker secrets are required at runtime.
 
 The following configuration values must be set via environment variables.
 
-| Variable Name                | Description                                                                                                                                                                                                                                               |
-| ---------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
-| `APP_URL`                    | The hostname of the server (e.g., `https://mycustomdomain.senecacollege.ca`). NOTE: when running in development, use `http://host.docker.internal:8080` vs. `http://localhost`, so that Docker DNS resolution works between the login container and host. |
-| `PORT`                       | The server runs on port `8080` by default                                                                                                                                                                                                                 |
-| `LOG_LEVEL`                  | The log level to use for log messages. One of `error`, `debug`, `info`, etc. See [Winston docs](https://github.com/winstonjs/winston#logging-levels). Defaults to `info`                                                                                  |
-| `ROOT_DOMAIN`                | The DNS root domain for the hosted zone (e.g., `starchart.com`)                                                                                                                                                                                           |
-| `AWS_ROUTE53_HOSTED_ZONE_ID` | The existing Amazon Route53 Hosted Zone ID to use (e.g., `Z23ABC4XYZL05B`)                                                                                                                                                                                |
-| `NOTIFICATIONS_EMAIL_USER`   | The email address from which notifications are sent                                                                                                                                                                                                       |
-| `SMTP_PORT`                  | The port to use for the SMTP server. Defaults to `587` in production (using `smtp.office365.com`) and `1025` in development (using ([MailHog](https://github.com/mailhog/MailHog))                                                                        |
-| `LETS_ENCRYPT_ACCOUNT_EMAIL` | The email address to use for the app's [single Let's Encrypt account](https://letsencrypt.org/docs/integration-guide/#one-account-or-many)                                                                                                                |
-| `REDIS_URL`                  | The Redis server to use for the worker queues. Defaults to `redis://redis:6379` in production and `localhost:6379` in development.                                                                                                                        |
-| `SAML_IDP_METADATA_PATH`     | The file path of the SAML Identify Provider (IdP)'s metadata XML. We store various XML files in `config/` and use `config/idp-metadata-dev.xml` by default in development.                                                                                |
-| `SECRETS_OVERRIDE`           | In development, to override the Docker secrets                                                                                                                                                                                                            |
-| `DATABASE_SETUP`             | In order to sync the Prisma schema with MySQL, set `DATABASE_SETUP=1` when running the app. NOTE: this **wipes all data** in MySQl and Redis, so be careful!                                                                                              |
+| Variable Name                | Description                                                                                                                                                                                                                                          |
+| ---------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `APP_URL`                    | The URL of the server (e.g., `https://mycustomdomain.senecacollege.ca`). NOTE: when running in development, use `http://host.docker.internal:8080` vs. `http://localhost`, so that Docker DNS resolution works between the login container and host. |
+| `PORT`                       | The server runs on port `8080` by default                                                                                                                                                                                                            |
+| `LOG_LEVEL`                  | The log level to use for log messages. One of `error`, `debug`, `info`, etc. See [Winston docs](https://github.com/winstonjs/winston#logging-levels). Defaults to `info`                                                                             |
+| `ROOT_DOMAIN`                | The DNS root domain for the hosted zone (e.g., `starchart.com`)                                                                                                                                                                                      |
+| `AWS_ROUTE53_HOSTED_ZONE_ID` | The existing Amazon Route53 Hosted Zone ID to use (e.g., `Z23ABC4XYZL05B`)                                                                                                                                                                           |
+| `NOTIFICATIONS_EMAIL_USER`   | The email address from which notifications are sent                                                                                                                                                                                                  |
+| `SMTP_PORT`                  | The port to use for the SMTP server. Defaults to `587` in production (using `smtp.office365.com`) and `1025` in development (using ([MailHog](https://github.com/mailhog/MailHog))                                                                   |
+| `LETS_ENCRYPT_ACCOUNT_EMAIL` | The email address to use for the app's [single Let's Encrypt account](https://letsencrypt.org/docs/integration-guide/#one-account-or-many)                                                                                                           |
+| `REDIS_URL`                  | The Redis server to use for the worker queues. Defaults to `redis://redis:6379` in production and `localhost:6379` in development.                                                                                                                   |
+| `SAML_IDP_METADATA_PATH`     | The file path of the SAML Identify Provider (IdP)'s metadata XML. We store various XML files in `config/` and use `config/idp-metadata-dev.xml` by default.                                                                                          |
+| `SECRETS_OVERRIDE`           | In development, to override the Docker secrets                                                                                                                                                                                                       |
+| `DATABASE_SETUP`             | In staging and production, use `DATABASE_SETUP=1` to run extra scripts on startup to create or sync the database with the Prisma schema. NOTE: this **wipes all data** in MySQl and Redis, so be careful!                                            |
 
 ### Secrets
 
@@ -51,9 +51,16 @@ The following secrets must be added to the Docker engine using [Docker Swarm sec
 The app can be started via Docker. If it is the first time you are starting the app, or if there are database changes that need to be applied first, you can set `DATABASE_SETUP=1` in the environment to run Prisma and sync the schema and database (NOTE: this **wipes all data** from MySQL and Redis so be careful!):
 
 ```sh
-# Run the app without doing anything to the database
-$ docker run ghcr.io/developingspace/starchart
-
-# OR, run the app, but sync the datbase first
-$ docker run -e DATABASE_SETUP=1 ghcr.io/developingspace/starchart
+# Run the app on staging (use -production for prod) without doing anything to the database
+$ docker stack deploy -c docker-staging.yml starchart
 ```
+
+To stop the app:
+
+```sh
+docker stack rm -c docker-staging.yml starchart
+```
+
+## Webhook Deployment Setup
+
+See the [Webhook Deploy docs in webhook/](webhook/README.md) for details about how to setup the deployment webhook.

--- a/webhook/README.md
+++ b/webhook/README.md
@@ -1,0 +1,34 @@
+# Starchart Deploy Webhook
+
+Starchart uses a GitHub Actions Workflow to automate deployment to staging and production. This is done using [webhook](https://github.com/adnanh/webhook).
+
+## Setup
+
+The `deploy.sh` file and `hooks.json` should be placed together, and `hooks.json` updated to include the correct paths:
+
+```json
+[
+  {
+    "id": "deploy",
+    "execute-command": "/path/deploy.sh",
+    "command-working-directory": "/path",
+```
+
+An HMAC secret needs to be installed in GitHub Actions Encrypted Secrets, as well as in the `hooks.json` file:
+
+```json
+"trigger-rule": {
+  "match": {
+    "type": "payload-hmac-sha256",
+    "secret": "TODO: update me to match secret stored in GitHub Actions...",
+    "parameter": {
+      "source": "header",
+      "name": "X-Hub-Signature"
+    }
+  }
+}
+```
+
+When the webhook is called, the secret is used to create a digest using the body, which can only be verified by using the same secret. The secret itself is never transmitted between GitHub and the deployment server.
+
+The `deploy.sh` script assumes that the Docker image for Starchart is at `ghcr.io/developingspace/starchart` and that the running Docker service is named `starchart_mycustomdomain`. If these are not correct, update the variables in `deploy.sh`.

--- a/webhook/README.md
+++ b/webhook/README.md
@@ -32,3 +32,33 @@ An HMAC secret needs to be installed in GitHub Actions Encrypted Secrets, as wel
 When the webhook is called, the secret is used to create a digest using the body, which can only be verified by using the same secret. The secret itself is never transmitted between GitHub and the deployment server.
 
 The `deploy.sh` script assumes that the Docker image for Starchart is at `ghcr.io/developingspace/starchart` and that the running Docker service is named `starchart_mycustomdomain`. If these are not correct, update the variables in `deploy.sh`.
+
+The `starchart-webhook.service` file defines a [systemd unit file](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_basic_system_settings/assembly_working-with-systemd-unit-files_configuring-basic-system-settings) to be installed on the server.
+
+Make sure that you update this file with the correct paths/ports for your `hooks.json` file, etc. Once it is configured properly, use the following to install and start it:
+
+```sh
+touch /etc/systemd/system/starchart-webhook.service
+chmod 664 /etc/systemd/system/starchart-webhook.service
+systemctl daemon-reload
+systemctl start starchart-webhook.service
+```
+
+> NOTE: updating this may require changes to SELinux
+
+The following commands can also be used to interact with the service:
+
+- `systemctl enable` - at next boot
+- `systemctl disable` - at next boot
+- `systemctl start` - now
+- `systemctl stop` - now
+- `systemctl enable` - `--now` combines enable and start
+- `systemctl disable` - `--now` combines disable and stop
+- `systemctl restart` - sequentially combines stop and start
+- `systemctl status` - show the status of the service
+
+To see the logs for the service, use:
+
+```sh
+sudo journalctl -u starchart-webhook
+```

--- a/webhook/deploy.sh
+++ b/webhook/deploy.sh
@@ -1,0 +1,22 @@
+#!/usr/bin/env bash
+set -eo pipefail
+
+# https://github.com/DevelopingSpace/starchart/pkgs/container/starchart
+IMAGE=ghcr.io/developingspace/starchart
+SERVICE=starchart_mycustomdomain
+
+# We expect to receive the new image tag via the TAG environment variable
+if [[ -z "$TAG" ]]; then
+    logger -s "webhook received, no tag set, skipping."
+    exit 1
+fi
+
+logger -s "webhook received, starting update of ${SERVICE} to ${IMAGE}:${TAG}"
+
+if docker service update --image "$IMAGE":"$TAG" "$SERVICE"; then
+    logger -s "webhook update of ${SERVICE} to ${IMAGE}:${TAG} succeeded."
+    exit 0
+else
+    logger -s "webhook update of ${SERVICE} to ${IMAGE}:${TAG} failed."
+    exit 1
+fi

--- a/webhook/hooks.json
+++ b/webhook/hooks.json
@@ -1,0 +1,24 @@
+[
+  {
+    "id": "deploy",
+    "execute-command": "/home/customdomain/deploy.sh",
+    "command-working-directory": "/home/customdomain",
+    "pass-environment-to-command": [
+      {
+        "source": "payload",
+        "name": "tag",
+        "envname": "TAG"
+      }
+    ],
+    "trigger-rule": {
+      "match": {
+        "type": "payload-hmac-sha256",
+        "secret": "TODO: update me to match secret stored in GitHub Actions...",
+        "parameter": {
+          "source": "header",
+          "name": "X-Hub-Signature"
+        }
+      }
+    }
+  }
+]

--- a/webhook/starchart-webhook.service
+++ b/webhook/starchart-webhook.service
@@ -1,0 +1,23 @@
+# systemd unit file for Starchart Webhook
+[Unit]
+Description=An automated deployment webhook for mycustomdomain (starchart)
+
+Documentation=https://github.com/DevelopingSpace/starchart
+
+# Start after the network is up
+After=network.target
+
+[Service]
+Type=simple
+
+# See https://github.com/adnanh/webhook 
+# Set the path to hooks.json and port to use
+ExecStart=webhook -hooks /path/to/hooks.json -port 8000
+# Specify these if want/need for the above to run
+#WorkingDirectory=
+#User=
+
+Restart=always
+
+[Install]
+WantedBy=multi-user.target

